onnx/shape_inference/implementation.cc

```diff
@@ -183,8 +183,34 @@ static void InferShapesImpl(
   }
 
   std::unordered_map<std::string, const TensorProto*> inputDataByName;
+  GraphProto inputGraph; // to not modify original graph; use temp graph  
   for (const auto& tp : g->initializer()) {
     inputDataByName[tp.name()] = &tp;
+    
+    // Support new IR: some tensors can only exist in initializer and not in input
+    // So shape_inference should make use of initializer shapes
+    // Store initializer shape info in value_info as well
+    TypeProto_Tensor* newTensorFromInitializer;
+    auto iter = valueTypesByName.find(tp.name());
+    // If it already exists in input, simply use the original input
+    if (iter != valueTypesByName.end()) {
+      continue; // skip if shape info exists in input 
```

Comment:
I think it would be useful, in this case, to check that the declared type/shape (from input lis) is compatible with the actual value in the initializer. I suggest creating a new "inferred" type (just like newTensorFromInitializer below) and calling checkShapesAndTypes on the original type and this new inferred type.

Reply:
the behavior varies from ir_version < 4 vs ir_version >=4
In case of ir_version < 4 the check which Rama suggested makes sense
In case of ir_version >=4 the shape from input should be prefer as the value in initializer is merely a default value

Reply:
A couple of comments: (a) Whatever the IR version, the type of the input-variable should be consistent with the value in the initializer. It is okay if the type of the input-variable has less precise-shape than the initializer, but it should be consistent. If the type of the input-variable is a float-tensor and the initializer is an int-tensor, it is incorrect; similarly, if the shape of the input-variable is [B, 10] while the initializer-value has shape [1, 5] it is incorrect. (b) If the input is present, we should use the shape given in the input-type, and not the more-precise shape from the initializer (even though we can check that they are consistent).